src/pages/instances/actions/snapshots/useInstanceSnapshot.tsx

```diff
@@ -0,0 +1,258 @@
+} from "util/helpers";
+import { getInstanceSnapshotSchema } from "util/instanceSnapshots";
+import { queryKeys } from "util/queryKeys";
+import { SnapshotFormValues, getExpiresAt } from "util/snapshots";
+import { TOOLTIP_OVER_MODAL_ZINDEX } from "util/zIndex";
+
+interface CreateInstanceSnapshotHook {
+  instance: LxdInstance;
+  close: () => void;
+  onSuccess: (message: ReactNode) => void;
+  isStateful: boolean;
+  isRunning: boolean;
+}
+
+interface EditInstanceSnapshotHook {
+  instance: LxdInstance;
+  snapshot: LxdInstanceSnapshot;
+  close: () => void;
+  onSuccess: (message: ReactNode) => void;
+}
+
+const useCreateInstanceSnapshot = (props: CreateInstanceSnapshotHook) => {
+  const { instance, close, onSuccess, isStateful, isRunning } = props;
+  const eventQueue = useEventQueue();
+  const notify = useNotify();
+  const queryClient = useQueryClient();
+  const controllerState = useState<AbortController | null>(null);
+
+  const formik = useFormik<SnapshotFormValues<{ stateful?: boolean }>>({
+    initialValues: {
+      name: "",
+      stateful: false,
+      expirationDate: null,
+      expirationTime: null,
+    },
+    validateOnMount: true,
+    validationSchema: getInstanceSnapshotSchema(instance, controllerState),
+    onSubmit: (values, { resetForm }) => {
+      notify.clear();
+      const expiresAt =
+        values.expirationDate && values.expirationTime
+          ? stringToIsoTime(
+              getExpiresAt(values.expirationDate, values.expirationTime),
+            )
+          : UNDEFINED_DATE;
+      void createInstanceSnapshot(
+        instance,
+        values.name,
+        expiresAt,
+        values.stateful || false,
+      )
+        .then((operation) =>
+          eventQueue.set(
+            operation.metadata.id,
+            () => {
+              void queryClient.invalidateQueries({
+                predicate: (query) => query.queryKey[0] === queryKeys.instances,
+              });
+              onSuccess(
+                <>
+                  Snapshot <ItemName item={values} bold /> created.
+                </>,
+              );
+              resetForm();
+              close();
+            },
+            (msg) => {
+              notify.failure("Snapshot creation failed", new Error(msg));
+              formik.setSubmitting(false);
+              close();
+            },
+          ),
+        )
+        .catch((error: Error) => {
+          notify.failure("Snapshot creation failed", error);
+          formik.setSubmitting(false);
+          close();
+        });
+    },
+  });
+
+  let statefulInfoMessage: JSX.Element | string = (
+    <>
+      {`To create a stateful snapshot, the instance needs\n`}
+      the <code>migration.stateful</code> config set to true
+    </>
+  );
+  if (isStateful) {
+    statefulInfoMessage = `To create a stateful snapshot,\nthe instance must be running`;
+  }
+  if (isStateful && isRunning) {
+    statefulInfoMessage = "";
+  }
+
+  const statefulCheckbox = (
```

Comment:
I'm going to move the hook logic into the CreateInstanceSnapshotForm. Since the useCreateInstanceSnapshot hook is used exclusively in this component now. The downside is that the component contains more logic, but the code style would be more consistent with the rest of the codebase. Same thing will be done for the other relevant hooks and components. wdyt?

Reply:
sounds good :+1: